src/components/DataForm/index.tsx

```diff
@@ -1,24 +1,43 @@
 "use client"
-import { DEFAULT_DATA } from "@/app/charts/constants";
-import { FC, Fragment, useState } from "react";
+
+import { FC, Fragment, useContext } from "react";
+import DataContext from "@/app/DataContext";
+import {BaseData} from "@/app/charts/types";
+
 
 const DataForm: FC<{}> = () => {
-  const [data, setData] = useState(DEFAULT_DATA);
-  return (
-    <>
+  const { data, setData } = useContext(DataContext);
+  const handleChange = (event: React.ChangeEvent<HTMLInputElement>, label: String, index: number) => {
+
+    setData({
+      ...data,
+      dataset:
+        data.dataset.map((d: BaseData, i:number) => {
+          if (i != index) {
+            return d;
+          }
+          let newData = {...d};
+          newData.data[label] = event.target.value;
+          return newData;
+        })
+    });
+  }
```

Comment:
So I love that you got your hands dirty with React Context here, it's typically something people find confusing initially 😅 

I will say though, we _typically_ try to avoid it unless if it's strictly necessary. As you probably saw, it's a bit complex to setup and it gets messy if you have a ton of context going on. You typically reserve it for a case where you're certain there's global state that needs to be shared, or if you're trying to pass data between grandparent and grandchildren components (2+ layers) because you don't want to be constantly chaining props.

In a case like this where the data only needs to flow between a single parent-child relationship, I would recommend doing it through props. You can split up props into two types, "inputs" and "outputs". In this case we need both because the data is flowing both ways. Inputs come in the form of basic variables, and outputs are callbacks (a function provided by the parent that will be called when the child needs to output something).

So in the case of here where we're trying to get an input of the current formData and an output of the modified formData when we make changes, we can setup the props for the component like this
```tsx
const DataForm: FC<{data: BaseData, onDataChange: (data: BaseData) => void}> = ({data, onDataChange}) => {
```
`data` is the input and `onDataChange` is the output.

Now we can also simplify the `set` function you did in here using the [set function from lodash](https://lodash.com/docs/4.17.15#set) from the [fp package](https://github.com/lodash/lodash/wiki/FP-Guide) (to keep it immutable). It would look something like this
```ts
set(`dataset[${index}].data.${label}`, Number(event.target.value), data)
```
Where the first arg is the path, the second arg is the value we're setting, and the final arg is the object we're copying, modifying and returning.

Putting this together, we can now remove your context from here and just have the component as
```tsx
const DataForm: FC<{data: BaseData, onDataChange: (data: BaseData) => void}> = ({data, onDataChange}) => {

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>, label: String, index: number) =>
    onDataChange(set(`dataset[${index}].data.${label}`, Number(event.target.value), data));
```

Then in `page.tsx` you're able to delete your context provider, and the node for `DataForm` will now look like this
```tsx
        <DataForm data={data} onDataChange={setData}/>
```

Now you just make each chart consume `data` instead of `DEFAULT_DATA` and you still have your dynamic charts, but without the Context 😄 

```tsx
<Bar
              title={data.title}
              data={createBarChartData(data, theme)}
              options={setBarChartOptions(data.title + " bar chart")}
            />
```

I totally understand that this was meant to just be a simple demo, so nothing was supposed to be perfect in here, but wanted to let you know how to do this for when we do start coding the app so that you don't feel like you need to use Context always. Let me know if this all makes sense or if you have any questions about it. Otherwise, this example looks great, awesome job! 🚀

Reply:
Thank you for the insight here. A lot of these things introduced on this demo we're actually quite confusing 😅 . My familiarity using React hooks as well as typescript is pretty much zero. 

I did see you used lodash for a few actions. I'll have to take a look at this library to see all the capabilities. 

I'm pretty stoked that I got it working in the first place, but you are correct with improving for when we start developing for the app. I'll see if I can refactor it in the way you suggestest.